static/js/containers/pages/LoginPage.js

```diff
@@ -0,0 +1,135 @@
+// @flow
+import React from "react"
+import { compose } from "redux"
+import { connect } from "react-redux"
+import { mutateAsync } from "redux-query"
+
+import queries from "../../lib/queries"
+import { routes, getNextParam } from "../../lib/urls"
+import { STATE_SUCCESS, STATE_LOGIN_PASSWORD } from "../../lib/auth"
+
+import { LoginEmailForm, LoginPasswordForm } from "../../components/forms/login"
+
+import type { RouterHistory, Location } from "react-router"
+import type { Response } from "redux-query"
+import type { AuthResponse } from "../../flow/authTypes"
+
+const STEP_EMAIL = "email"
+const STEP_PASSWORD = "password"
+
+type Props = {
+  location: Location,
+  history: RouterHistory,
+  loginEmail: (email: string, next: ?string) => Promise<Response<AuthResponse>>,
+  loginPassword: (
+    password: string,
+    partialToken: string
+  ) => Promise<Response<AuthResponse>>
+}
+
+type State = {
+  step: "email" | "password",
+  partialToken: ?string,
+  name: ?string
+}
+
+class LoginPage extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props)
+    this.state = {
+      step:         STEP_EMAIL,
+      partialToken: null,
+      name:         null
+    }
+  }
+
+  async onSubmitEmail({ email }, { setSubmitting, setErrors }) {
+    const { loginEmail, location } = this.props
+    const nextUrl = getNextParam(location.search)
+
+    /* eslint-disable camelcase */
+    try {
+      const {
+        body: { state, partial_token, errors, extra_data }
+      }: { body: AuthResponse } = await loginEmail(email, nextUrl)
+
+      if (state === STATE_LOGIN_PASSWORD) {
+        this.setState({
+          step:         STEP_PASSWORD,
+          partialToken: partial_token,
+          name:         extra_data.name
+        })
+      } else if (errors.length > 0) {
+        setErrors({
+          email: errors[0]
+        })
+      }
+    } finally {
+      setSubmitting(false)
+    }
+    /* eslint-enable camelcase */
+  }
+
+  async onSubmitPassword({ password }, { setSubmitting, setErrors }) {
+    const { loginPassword, history } = this.props
+    const { partialToken } = this.state
+
+    if (!partialToken) {
+      throw Error("Invalid state: password page with no partialToken")
+    }
+
+    /* eslint-disable camelcase */
+    try {
+      const {
+        body: { state, redirect_url, errors }
+      }: { body: AuthResponse } = await loginPassword(password, partialToken)
+
+      if (state === STATE_SUCCESS) {
+        history.push(redirect_url || routes.home)
+      } else if (errors.length > 0) {
```

Comment:
If I enter a password that is less than 8 chars, the response from the server is a 400 with this in the body: `{"password":["Ensure this field has at least 8 characters."]}`. That means no `errors` key, so the error isn't shown in the form. If I had to guess, this will probably involve a change to `authentication.serializers.LoginPasswordSerializer`.

Reply:
Hmm, I'll have to look into that, I suppose for now I can just add it to the frontend validation, which is probably the preferable and more user-friendly way to do it anyway.